File: local-marketplace/plugins/claude-exo-self/hooks-handlers/extract-handoff.py
#!/usr/bin/env -S uv run python
"""Extract a structured handoff summary from a Claude Code transcript.

Called by pre-compact.sh to automatically save session state before compaction.
Reads the transcript JSONL, extracts user prompts, files modified, and the
last assistant response to produce a compact summary for the next instance.
"""
import json
import sys


def extract_handoff(transcript_path, max_chars=3000):
    """Parse transcript and produce a structured handoff summary."""
    user_prompts = []
    assistant_texts = []
    tools_used = set()
    files_modified = set()

    with open(transcript_path) as f:
        for line in f:
            try:
                obj = json.loads(line)
            except json.JSONDecodeError:
                continue

            if "message" not in obj:
                continue

            msg = obj["message"]
            role = msg.get("role")
            content = msg.get("content", "")

            if role == "user":
                if isinstance(content, str) and content.strip():
                    text = content.strip()
                    if not text.startswith("<system-reminder>"):
                        user_prompts.append(text[:200])
                elif isinstance(content, list):
                    for block in content:
                        if isinstance(block, dict) and block.get("type") == "text":
                            text = block.get("text", "").strip()
                            if text and not text.startswith("<system-reminder>"):
                                user_prompts.append(text[:200])

            elif role == "assistant":
                if isinstance(content, list):
                    for block in content:
                        if not isinstance(block, dict):
                            continue
                        btype = block.get("type")
                        if btype == "text":
                            text = block.get("text", "").strip()
                            if text:
                                assistant_texts.append(text)
                        elif btype == "tool_use":
                            tool_name = block.get("name", "")
                            tools_used.add(tool_name)
                            tool_input = block.get("input", {})
                            if tool_name in ("Edit", "Write"):
                                fp = tool_input.get("file_path", "")
                                if fp:
                                    files_modified.add(fp)

    sections = []

    # What the user asked for
    if user_prompts:
        first = user_prompts[:3]
        last = user_prompts[-2:] if len(user_prompts) > 3 else []
        prompt_summary = "\n".join(f"- {p}" for p in first)
        if last:
            prompt_summary += "\n...\n" + "\n".join(f"- {p}" for p in last)
        sections.append(f"## User Requests\n\n{prompt_summary}")

    if files_modified:
        files_list = "\n".join(f"- {f}" for f in sorted(files_modified))
        sections.append(f"## Files Modified\n\n{files_list}")

    if tools_used:
        sections.append(f"## Tools Used\n\n{', '.join(sorted(tools_used))}")

    if assistant_texts:
        last_response = assistant_texts[-1]
        if len(last_response) > 800:
            last_response = last_response[:800] + "..."
        sections.append(f"## Last Response Summary\n\n{last_response}")

    result = "\n\n".join(sections)
    if len(result) > max_chars:
        result = result[:max_chars - 3] + "..."

    return result


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: extract-handoff.py <transcript.jsonl>", file=sys.stderr)
        sys.exit(1)
    print(extract_handoff(sys.argv[1]))

File: plugins/claude-exo-self/src/commands/extract_handoff.rs
use serde_json::Value;
use std::collections::BTreeSet;
use std::io::BufRead;

/// Extract a structured handoff summary from a Claude Code transcript JSONL.
/// Called by pre-compact to automatically save session state before compaction.
pub fn run(transcript_path: &str) {
    let result = extract(transcript_path, 3000);
    print!("{result}");
}

fn extract(transcript_path: &str, max_chars: usize) -> String {
    let file = match std::fs::File::open(transcript_path) {
        Ok(f) => f,
        Err(_) => return String::new(),
    };

    let reader = std::io::BufReader::new(file);
    let mut user_prompts = Vec::new();
    let mut assistant_texts = Vec::new();
    let mut tools_used = BTreeSet::new();
    let mut files_modified = BTreeSet::new();

    for line in reader.lines() {
        let Ok(line) = line else { continue };
        let Ok(obj) = serde_json::from_str::<Value>(&line) else {
            continue;
        };

        let Some(msg) = obj.get("message") else {
            continue;
        };
        let role = msg.get("role").and_then(|r| r.as_str()).unwrap_or("");
        let content = &msg["content"];

        match role {
            "user" => {
                if let Some(text) = content.as_str() {
                    let text = text.trim();
                    if !text.is_empty() && !text.starts_with("<system-reminder>") {
                        user_prompts.push(truncate(text, 200));
                    }
                } else if let Some(arr) = content.as_array() {
                    for block in arr {
                        if block.get("type").and_then(|t| t.as_str()) == Some("text") {
                            let text = block
                                .get("text")
                                .and_then(|t| t.as_str())
                                .unwrap_or("")
                                .trim();
                            if !text.is_empty() && !text.starts_with("<system-reminder>") {
                                user_prompts.push(truncate(text, 200));
                            }
                        }
                    }
                }
            }
            "assistant" => {
                if let Some(arr) = content.as_array() {
                    for block in arr {
                        let btype = block.get("type").and_then(|t| t.as_str()).unwrap_or("");
                        match btype {
                            "text" => {
                                let text = block
                                    .get("text")
                                    .and_then(|t| t.as_str())
                                    .unwrap_or("")
                                    .trim();
                                if !text.is_empty() {
                                    assistant_texts.push(text.to_string());
                                }
                            }
                            "tool_use" => {
                                let tool_name =
                                    block.get("name").and_then(|n| n.as_str()).unwrap_or("");
                                if !tool_name.is_empty() {
                                    tools_used.insert(tool_name.to_string());
                                }
                                if (tool_name == "Edit" || tool_name == "Write")
                                    && let Some(fp) = block
                                        .get("input")
                                        .and_then(|i| i.get("file_path"))
                                        .and_then(|p| p.as_str())
                                    {
                                        files_modified.insert(fp.to_string());
                                    }
                            }
                            _ => {}
                        }
                    }
                }
            }
            _ => {}
        }
    }

    let mut sections = Vec::new();

    // User requests
    if !user_prompts.is_empty() {
        let first: Vec<_> = user_prompts.iter().take(3).collect();
        let mut prompt_summary: String = first
            .iter()
            .map(|p| format!("- {p}"))
            .collect::<Vec<_>>()
            .join("\n");
        if user_prompts.len() > 3 {
            let last: Vec<_> = user_prompts
                .iter()
                .rev()
                .take(2)
                .collect::<Vec<_>>()
                .into_iter()
                .rev()
                .collect();
            prompt_summary.push_str("\n...\n");
            prompt_summary.push_str(
                &last
                    .iter()
                    .map(|p| format!("- {p}"))
                    .collect::<Vec<_>>()
                    .join("\n"),
            );
        }
        sections.push(format!("## User Requests\n\n{prompt_summary}"));
    }

    if !files_modified.is_empty() {
        let list = files_modified
            .iter()
            .map(|f| format!("- {f}"))
            .collect::<Vec<_>>()
            .join("\n");
        sections.push(format!("## Files Modified\n\n{list}"));
    }

    if !tools_used.is_empty() {
        let list = tools_used.into_iter().collect::<Vec<_>>().join(", ");
        sections.push(format!("## Tools Used\n\n{list}"));
    }

    if let Some(last_response) = assistant_texts.last() {
        let mut summary = last_response.clone();
        if summary.len() > 800 {
            summary.truncate(800);
            summary.push_str("...");
        }
        sections.push(format!("## Last Response Summary\n\n{summary}"));
    }

    let mut result = sections.join("\n\n");
    if result.len() > max_chars {
        result.truncate(max_chars - 3);
        result.push_str("...");
    }
    result
}

fn truncate(s: &str, max: usize) -> String {
    if s.len() <= max {
        s.to_string()
    } else {
        format!("{}...", &s[..max - 3])
    }
}
